### PR TITLE
Refactor the qf_parse_line() function in quickfix.c

### PR DESCRIPTION
**Pull request by yegappan**
The qf_parse_line() function in quickfix.c is too long. Refactor the function.
Add additional comments to some of the functions. No new functionality
is introduced in this patch.

### CODE REVIEW

File: src/quickfix.c [hunk 1]
```diff
@@ -824,6 +825,349 @@ typedef struct {
...
+	if (regmatch->startp[i] == NULL)
+	    return QF_FAIL;
+	fields->type = *regmatch->startp[i];
+    }
+    if (fmt_ptr->flags == '+' && !qf_multiscan)	/* %+ */
+    {
+	if (linelen >= fields->errmsglen)
+	{
+	    /* linelen + null terminator */
+	    if ((fields->errmsg = vim_realloc(fields->errmsg,
```
Comment (dpelle) [inline src/quickfix.c]: This way of calling realloc leaks memory if realloc(…) fails i.e. if it returns NULL.
See https://www.viva64.com/en/w/v701/ for more details.

File: src/quickfix.c [hunk 2]
```diff
@@ -824,6 +825,349 @@ typedef struct {
...
+    }
+    else if ((i = (int)fmt_ptr->addr[5]) > 0)	/* %m */
+    {
+	if (regmatch->startp[i] == NULL || regmatch->endp[i] == NULL)
+	    return QF_FAIL;
+	len = (int)(regmatch->endp[i] - regmatch->startp[i]);
+	if (len >= fields->errmsglen)
+	{
+	    /* len + null terminator */
+	    if ((fields->errmsg = vim_realloc(fields->errmsg, len + 1))
```
Comment (dpelle) [inline src/quickfix.c]: Same remark: leak if vim_realloc(…) returns NULL.

File: src/quickfix.c [hunk 3]
```diff
@@ -824,6 +825,349 @@ typedef struct {
...
+    static int
+qf_parse_line_nomatch(char_u *linebuf, int linelen, qffields_T *fields)
+{
+    fields->namebuf[0] = NUL;	/* no match found, remove file name */
+    fields->lnum = 0;			/* don't jump to this line */
+    fields->valid = FALSE;
+    if (linelen >= fields->errmsglen)
+    {
+	/* linelen + null terminator */
+	if ((fields->errmsg = vim_realloc(fields->errmsg,
```
Comment (dpelle) [inline src/quickfix.c]: Leak if vim_realloc(…) fails.

**Comment by vim-ml** [comment]
Hi,

On Mon, May 7, 2018 at 12:12 AM, Dominique Pellé
<vim-dev-github@256bit.org> wrote:
> @dpelle commented on this pull request.
>
> ________________________________
>
> In src/quickfix.c:
>
>> + if (regmatch->startp[i] == NULL)
> +    return QF_FAIL;
> + fields->col = (int)atol((char *)regmatch->startp[i]);
> +    }
> +    if ((i = (int)fmt_ptr->addr[4]) > 0) /* %t */
> +    {
> + if (regmatch->startp[i] == NULL)
> +    return QF_FAIL;
> + fields->type = *regmatch->startp[i];
> +    }
> +    if (fmt_ptr->flags == '+' && !qf_multiscan) /* %+ */
> +    {
> + if (linelen >= fields->errmsglen)
> + {
> +    /* linelen + null terminator */
> +    if ((fields->errmsg = vim_realloc(fields->errmsg,
>
> This way of calling realloc leaks memory if realloc(…) fails i.e. if it
> returns NULL.
> See https://www.viva64.com/en/w/v701/ for more details.
>

Thanks for pointing this out. I have pushed out a new commit that
fixes this issue. I have also fixed other places in quickfix.c where
vim_realloc() is used.

Thanks,
Yegappan

>
> ________________________________
>
> In src/quickfix.c:
>
>> +    if ((fields->errmsg = vim_realloc(fields->errmsg,
> +    linelen + 1)) == NULL)
> + return QF_NOMEM;
> +    fields->errmsglen = linelen + 1;
> + }
> + vim_strncpy(fields->errmsg, linebuf, linelen);
> +    }
> +    else if ((i = (int)fmt_ptr->addr[5]) > 0) /* %m */
> +    {
> + if (regmatch->startp[i] == NULL || regmatch->endp[i] == NULL)
> +    return QF_FAIL;
> + len = (int)(regmatch->endp[i] - regmatch->startp[i]);
> + if (len >= fields->errmsglen)
> + {
> +    /* len + null terminator */
> +    if ((fields->errmsg = vim_realloc(fields->errmsg, len + 1))
>
> Same remark: leak if vim_realloc(…) returns NULL.
>
> ________________________________
>
> In src/quickfix.c:
>
>> +}
> +
> +/*
> + * Parse a non-error line (a line which doesn't match any of the error
> + * format in 'efm').
> + */
> +    static int
> +qf_parse_line_nomatch(char_u *linebuf, int linelen, qffields_T *fields)
> +{
> +    fields->namebuf[0] = NUL; /* no match found, remove file name */
> +    fields->lnum = 0; /* don't jump to this line */
> +    fields->valid = FALSE;
> +    if (linelen >= fields->errmsglen)
> +    {
> + /* linelen + null terminator */
> + if ((fields->errmsg = vim_realloc(fields->errmsg,
>
> Leak if vim_realloc(…) fails.
>

**Pull request closed without merging**